romancal/tweakreg/tweakreg_step.py

```diff
@@ -103,52 +100,39 @@ def process(self, input):
-                        member["expname"] = os.path.join(asn_dir, filename)
-                        if filename in catdict:
-                            member["tweakreg_catalog"] = catdict[filename]
-                        elif "tweakreg_catalog" in member:
-                            del member["tweakreg_catalog"]
-
-                    images.from_asn(asn_data)
-                elif is_association(input):
-                    images.from_asn(input)
-                else:
-                    images = ModelContainer(input)
-                    for im in images:
-                        filename = im.meta.filename
-                        if filename in catdict:
-                            self.log.info(
-                                f"setting "
-                                f"{filename}.source_detection.tweakreg_catalog_name ="
-                                f" {repr(catdict[filename])}"
-                            )
-                            # set catalog name only (no catalog data at this point)
-                            im.meta["source_detection"] = {
-                                "tweakreg_catalog_name": catdict[filename],
-                            }
+            if isinstance(input, rdm.DataModel):
+                images = ModelLibrary([input])
+            elif str(input).endswith(".asdf"):
+                images = ModelLibrary([rdm.open(input)])
+            elif isinstance(input, ModelLibrary):
+                images = input
             else:
-                images = (
-                    ModelContainer([input])
-                    if (
-                        isinstance(input, rdm.DataModel) or str(input).endswith(".asdf")
-                    )
-                    else ModelContainer(input)
-                )
+                images = ModelLibrary(input)
         except TypeError as e:
             e.args = (
                 "Input to tweakreg must be a list of DataModels, an "
-                "association, or an already open ModelContainer "
+                "association, or an already open ModelLibrary "
                 "containing one or more DataModels.",
             ) + e.args[1:]
             raise e
 
+        if use_custom_catalogs and catdict:
+            with images:
+                for i, member in enumerate(images.asn["products"][0]["members"]):
+                    filename = member["expname"]
+                    if filename in catdict:
+                        # FIXME: I'm not sure if this captures all the possible combinations
+                        # for example, meta.tweakreg_catalog is set by the container (when
+                        # it's present in the association). However the code in this step
+                        # checks meta.source_catalog.tweakreg_catalog. I think this means
+                        # that setting a catalog via an association does not work. Is this
+                        # intended? If so, the container can be updated to not support that.
```

Comment:
This will change a meaningful amount when we stop storing the source detection catalog in the image metadata.  When that happens, tweakreg will need to get a list of images and associated catalogs.  I suspect that means that a lot of this logic could be simplified (e.g., all catalogs will be specified by a file name, and tweakreg doesn't need special logic to separate user catalogs from normal source catalogs).

Reply:
Thanks! That's very helpful to know. I am curious to hear more as the ability to set the catalog from the association, the model metadata and the "catfile" has led to [some confusion](https://github.com/spacetelescope/jwst/pull/8424#discussion_r1571302034) (and bugs) in jwst.